File: client/src/protocol.cc
#include "protocol.h"
#include <cstring>
#include <stdexcept>

std::vector<uint8_t> Protocol::packRequestHeader(
    const uint8_t* client_id,
    uint16_t code,
    uint32_t payload_size
) {
    std::vector<uint8_t> header(HEADER_SIZE);
    
    std::memcpy(header.data(), client_id, CLIENT_ID_SIZE);
    
    header[CLIENT_ID_SIZE] = VERSION;
    
    header[CLIENT_ID_SIZE + 1] = code & 0xFF;
    header[CLIENT_ID_SIZE + 2] = (code >> 8) & 0xFF;
    
    header[CLIENT_ID_SIZE + 3] = payload_size & 0xFF;
    header[CLIENT_ID_SIZE + 4] = (payload_size >> 8) & 0xFF;
    header[CLIENT_ID_SIZE + 5] = (payload_size >> 16) & 0xFF;
    header[CLIENT_ID_SIZE + 6] = (payload_size >> 24) & 0xFF;
    
    return header;
}

ResponseHeader Protocol::unpackResponseHeader(const std::vector<uint8_t>& data) {
    if (data.size() < 7) {
        throw std::runtime_error("Invalid response header size");
    }
    
    ResponseHeader header;
    header.version = data[0];
    header.code = data[1] | (data[2] << 8);
    header.payload_size = data[3] | (data[4] << 8) | (data[5] << 16) | (data[6] << 24);
    
    return header;
}

std::vector<uint8_t> Protocol::packRegisterRequest(
    const std::string& username,
    const std::vector<uint8_t>& public_key
) {
    uint8_t empty_id[CLIENT_ID_SIZE] = {0};
    uint32_t payload_size = USERNAME_MAX_SIZE + PUBLIC_KEY_SIZE;
    
    auto request = packRequestHeader(empty_id, REQ_REGISTER, payload_size);
    
    std::vector<uint8_t> username_bytes(USERNAME_MAX_SIZE, 0);
    size_t len = std::min(username.length(), size_t(USERNAME_MAX_SIZE - 1));
    std::memcpy(username_bytes.data(), username.c_str(), len);
    request.insert(request.end(), username_bytes.begin(), username_bytes.end());
    
    if (public_key.size() != PUBLIC_KEY_SIZE) {
        throw std::runtime_error("Invalid public key size");
    }
    request.insert(request.end(), public_key.begin(), public_key.end());
    
    return request;
}

std::vector<uint8_t> Protocol::packClientListRequest(const uint8_t* client_id) {
    return packRequestHeader(client_id, REQ_CLIENT_LIST, 0);
}

std::vector<uint8_t> Protocol::packPublicKeyRequest(
    const uint8_t* client_id,
    const uint8_t* target_client_id
) {
    auto request = packRequestHeader(client_id, REQ_PUBLIC_KEY, CLIENT_ID_SIZE);
    
    for (size_t i = 0; i < CLIENT_ID_SIZE; i++) {
        request.push_back(target_client_id[i]);
    }
    
    return request;
}

std::vector<uint8_t> Protocol::packWaitingMessagesRequest(const uint8_t* client_id) {
    return packRequestHeader(client_id, REQ_WAITING_MESSAGES, 0);
}

std::vector<uint8_t> Protocol::packSendMessageRequest(
    const uint8_t* from_client_id,
    const uint8_t* to_client_id,
    uint8_t msg_type,
    const std::vector<uint8_t>& content
) {
    uint32_t payload_size = CLIENT_ID_SIZE + 1 + 4 + content.size();
    auto request = packRequestHeader(from_client_id, REQ_SEND_MESSAGE, payload_size);
    
    for (size_t i = 0; i < CLIENT_ID_SIZE; i++) {
        request.push_back(to_client_id[i]);
    }
    
    request.push_back(msg_type);
    
    uint32_t content_size = content.size();
    request.push_back(content_size & 0xFF);
    request.push_back((content_size >> 8) & 0xFF);
    request.push_back((content_size >> 16) & 0xFF);
    request.push_back((content_size >> 24) & 0xFF);
    
    request.insert(request.end(), content.begin(), content.end());
    
    return request;
}

File: src/server/protocol.py
# Request codes
REQ_REGISTER = 600
REQ_CLIENT_LIST = 601
REQ_PUBLIC_KEY = 602
REQ_SEND_MESSAGE = 603
REQ_WAITING_MESSAGES = 604
REQ_EXIT = 0

# Response codes
RES_REGISTRATION_SUCCESS = 2100
RES_CLIENT_LIST = 2101
RES_PUBLIC_KEY = 2102
RES_MESSAGE_SENT = 2103
RES_WAITING_MESSAGES = 2104
RES_GENERAL_ERROR = 9000

# Message types
MSG_TYPE_SYM_KEY_REQUEST = 1
MSG_TYPE_SYM_KEY_SEND = 2
MSG_TYPE_TEXT_MESSAGE = 3
MSG_TYPE_FILE = 4

VERSION = 2
HEADER_SIZE = 23
CLIENT_ID_SIZE = 16
UUID_SIZE = 16
USERNAME_MAX_SIZE = 255
PUBLIC_KEY_SIZE = 160

class ProtocolError(Exception):
    pass

def pack_header(version, code, payload_size):
    import struct
    return struct.pack('<BHI', version, code, payload_size)

def unpack_request_header(data):
    import struct
    if len(data) < HEADER_SIZE:
        raise ProtocolError("Invalid header size")
    
    client_id = data[:CLIENT_ID_SIZE]
    version, code, payload_size = struct.unpack('<BHI', data[CLIENT_ID_SIZE:HEADER_SIZE])
    
    return {
        'client_id': client_id,
        'version': version,
        'code': code,
        'payload_size': payload_size
    }

def pack_response(code, payload):
    header = pack_header(VERSION, code, len(payload))
    return header + payload

def pack_client_info(client_id, name):
    import struct
    name_bytes = name.encode('ascii')[:254]
    name_bytes += b'\x00' * (USERNAME_MAX_SIZE - len(name_bytes))
    return client_id + name_bytes

def pack_message_info(client_id, message_id, msg_type, content):
    import struct
    content_size = len(content)
    header = client_id + struct.pack('<IBI', message_id, msg_type, content_size)
    return header + content

